hta/analyzers/critical_path_analysis.py

```diff
@@ -501,19 +606,33 @@ def handle_cuda_sync(row):
                 # Add launch delay edge if there were no outstanding kernels
                 # when the CUDA runtime was launching it.
                 runtime_start, _ = self.get_nodes_for_event(runtime_index)
-                assert runtime_start is not None
+                assert runtime_start is not None, (
+                    f"Could not find runtime index = {runtime_index}, current kernel "
+                    f"stream = {stream}, "
+                    f"name = {self._get_node_name(eid)}, correlation = {row['correlation']}"
+                )
                 self._add_edge_helper(
                     runtime_start, start_node, type=CPEdgeType.KERNEL_LAUNCH_DELAY
                 )
-            elif last_node.get(stream) is not None and kernel_sync_index is None:
+                edge_added = True
+            elif (
+                last_node.get(stream) is not None
+                # and the kernel sync dependency if any finished earlier than last node
+                and (
+                    kernel_sync_index is None
+                    or kernel_sync_end.ts < last_node[stream].ts
+                )
+            ):
                 # If neither launch nor CUDA event sync occurs it is a kernel-kernel
                 # delay
                 self._add_edge_helper(
                     last_node[stream], start_node, type=CPEdgeType.KERNEL_KERNEL_DELAY
                 )
-            elif kernel_sync_index is None:
+                edge_added = True
+            elif not edge_added:
+                # Neither of Sync, kernel-kernel or kernel launch edges were added
                 logger.error(
-                    f"Queue length is {queue_length_runtime}!= 1 but no "
+                    f"Final fall through Queue length is {queue_length_runtime}!= 1 but no "
```

Comment:
what does "Final fall through queue length" mean?

Reply:
Oh my bad, this means that it reached to the last fall through in the trace and it should not have gotten here. I'll improve the message.